photon-client/src/main/scala/com/linkedin/photon/ml/hyperparameter/ShrinkSearchRange.scala

```diff
@@ -0,0 +1,126 @@
+  */
+object ShrinkSearchRange {
+
+  /**
+    * Compute the lower bound and upper bound of the new prior config
+    *
+    * @param hyperParams Configurations of hyper-parameters
+    * @param priorJsonString JSON string containing prior observations
+    * @param priorDefault Default values for missing hyper-parameters
+    * @param radius The radius of search range after transformation to [0, 1]
+    * @return A tuple of lower bounds and upper bounds for each hyperparameter. Lower bounds and upper bounds are
+    *         discrete for discrete parameters. Lower bounds and upper bounds are dense vectors.
+    */
+  def getBounds(hyperParams: HyperparameterConfig,
+                priorJsonString: String,
+                priorDefault: Map[String, String],
+                radius: Double,
+                seed: Long = System.currentTimeMillis): (DenseVector[Double], DenseVector[Double]) = {
+
+    val hyperparameterList = hyperParams.names
+    val ranges = hyperParams.ranges
+    val discreteParams = hyperParams.discreteParams
+    val numParams = ranges.length
+
+    // Get a [[Seq]] of (vectorized hyper-parameters, evaluationValue) tuples
+    val hyperparameterPairs = HyperparameterSerialization.priorFromJson(priorJsonString, priorDefault, hyperparameterList)
+
+    // Rescale the hyperparameters to [0,1]
+    val hyperparameterRescaled = VectorRescaling.rescalePriors(hyperparameterPairs, hyperParams)
+
+    // Combine hyperparameters as a dense matrix and evaluation value as a dense vector
+    val (overallPoints, overallEvals) = hyperparameterRescaled.map(x => (x._1.asDenseMatrix, DenseVector(x._2))).reduce(
+      (a, b) => (DenseMatrix.vertcat(a._1, b._1), DenseVector.vertcat(a._2, b._2))
+    )
+
+    // Fit Gaussian process regression model
+    val estimator = new GaussianProcessEstimator(kernel = new Matern52)
+    val model = estimator.fit(overallPoints, overallEvals)
+
+    // Sobol generator
+    val paramDistributions = {
+      val sobol = new SobolSequenceGenerator(numParams)
+      sobol.skipTo((seed % (Int.MaxValue.toLong + 1)).toInt)
+      sobol
+    }
+
+    // Draw candidates from a Sobol generator
+    val candidates = (1 until 1000).foldLeft(DenseMatrix(paramDistributions.nextVector)) { case (acc, _) =>
+      DenseMatrix.vertcat(acc, DenseMatrix(paramDistributions.nextVector))
+    }
+
+    // Select the best candidate
+    val predictions = model.predict(candidates)
+    val bestCandidate = selectBestCandidate(candidates, predictions._1)
+
+    // compute lower bound and upper bound
+    val upperBound = VectorRescaling.scaleBackward(
+      discretizeCandidate(bestCandidate.map(x => min(x + radius, 1.0)), discreteParams),
+      ranges,
+      discreteParams.keySet)
+    val lowerBound = VectorRescaling.scaleBackward(
+      discretizeCandidate(bestCandidate.map(x => max(x - radius, 0.0)), discreteParams),
+      ranges,
+      discreteParams.keySet)
+
+    discreteParams.foreach{
```

Comment:
Should we always +- 1? I guess it might be a better idea to work with the radius on this domain as well. In the case we get only one integer then we expand that a bit. For example shrinking [4, 10] to [0,1] with 0.01 radius will usually give a single integer where we can extend. But [1, 10000] to [0,1] with 0.01 will contain more than 1 integer.

Reply:
Right now in photon-ml case class DoubleRange has the requirement that start is strictly less than end, therefore I manually expand the range. Should we change the case class requirements to be start <= end?